src/flatten.jl

```diff
@@ -0,0 +1,163 @@
+
+isassociative(x::NamedTuple) = true
+isassociative(x::AbstractDict) = true
+isassociative(x::Iterators.Pairs) = true
+isassociative(x::Vector{<:Pair}) = true
+isassociative(x) = false
+
+"""
+    flatten(collection, [delim])
+
+Flatten a collection of nested associative types into a flat collection of pairs.
+If the input keys are symbols (ie: `NamedTuple`) then the $DEFAULT_FLATTEN_DELIM  will be
+used, otherwise `Tuple` keys will be returned.
+
+# Example
+```jldoctest
+julia> using AxisSets: flatten
+
+julia> data = (
+           val1 = (a1 = 1, a2 = 2),
+           val2 = (b1 = 11, b2 = 22),
+           val3 = [111, 222],
+           val4 = 4.3,
+       );
+
+julia> flatten(data)
+(val1⁻a1 = 1, val1⁻a2 = 2, val2⁻b1 = 11, val2⁻b2 = 22, val3 = [111, 222], val4 = 4.3)
+```
+
+    flatten(A, dims, [delim])
+
+Flatten a `KeyedArray` along the specified consecutive dimensions.
+The `dims` argument can either be a `Tuple` of symbols or a `Pair{Tuple, Symbol}` if
+you'd like to specify the desired flattened dimension name.
+If the `dims` is just a `Tuple` with no output dimension specified then $DEFAULT_PROD_DELIM
+will be used to generate the new dimension name.
+
+# Example
+```jldoctest
+julia> using AxisKeys, Dates, NamedDims; using AxisSets: flatten
+
+julia> A = KeyedArray(
+           reshape(1:24, (4, 3, 2));
+           time=DateTime(2021, 1, 1, 11):Hour(1):DateTime(2021, 1, 1, 14),
+           obj=[:a, :b, :c],
+           loc=[1, 2],
+       );
+
+julia> dimnames(flatten(A, (:obj, :loc)))
+(:time, :objᵡloc)
+```
+"""
+function flatten end
+
+# Single arg version defaults to `nothing` delim
+# All other methods dispatch on the two arg form to be explicit
+flatten(x) = flatten(x, nothing)
+
+# Primary flatten algorithm. Other methods should call this and adjust keys as necessary.
+function flatten(x::Union{Vector{<:Pair}, Iterators.Pairs}, delim::Nothing)
+    result = Channel{Pair}() do chnl
+        for (k, v) in x
+            if isassociative(v)
+                children = flatten(v)
+                iter = isa(children, NamedTuple) ? pairs(children) : children
+                for (_k, _v) in iter
+                    new_key = isa(_k, Tuple) ? (k, _k...) : (k, _k)
+                    put!(chnl, new_key => _v)
+                end
+            else
+                put!(chnl, (k,) => v)
+            end
+        end
+    end
+
+    return collect(result)
+end
+
+# NOTE: NamedTuples only support symbol names, so we use a simple :⁻ delimiter
+function flatten(x::NamedTuple, delim::Symbol=DEFAULT_FLATTEN_DELIM)::NamedTuple
+    kwargs = map(flatten(pairs(x))) do (k, v)
+        _k = isa(k, Tuple) ? join(k, delim) : k
+        return Symbol(_k) => v
+    end
+    return (; kwargs...)
+end
+
+# Since the key type could be changed as part of flattening we can only ensure the that
+# result is a `Dict` (vs the same input dict).
+flatten(x::AbstractDict, delim) = Dict(flatten(collect(x), delim)...)
```

Comment:
do you want to restrict the key type to Symbol/String? otherwise we should test on other types if they are supported

Reply:
No, we don't want to restrict the key type here. I can add a few simple test, but this will likely reduce to testing the stringification and joinability of various types.

Reply:
Turns out a different method was actually being too restrictive, so adding a couple extra tests was helpful 👍🏻